### PR TITLE
Using SearchValues in ContentDispositionHeaderValue

### LINKED ISSUE #46484: Use SearchValues throughout dotnet/aspnetcore
Akin to https://github.com/dotnet/runtime/issues/78204

Some uses have already been replaced in #45300, but there are more cases where using `IndexOfAnyValues` can speed things up or simplify the implementation. From a few grep searches:
- [x]  [`PathString.ToUriComponent`](https://github.com/dotnet/aspnetcore/blob/83237c9afaa49a9cb90ceb4da56dddb744baeed0/src/Http/Http.Abstractions/src/PathString.cs#L67) can be rewritten along the lines of [`UriHelper.EscapeString`](https://github.com/dotnet/runtime/blob/167112574ebfcfcddfc23267e1ba93eebdcdeaf1/src/libraries/System.Private.Uri/src/System/UriHelper.cs#L110-L219)
- [x]  https://github.com/dotnet/aspnetcore/blob/83237c9afaa49a9cb90ceb4da56dddb744baeed0/src/Http/Http.Abstractions/src/HostString.cs#L120
- [x]  ~~https://github.com/dotnet/aspnetcore/blob/83237c9afaa49a9cb90ceb4da56dddb744baeed0/src/FileProviders/Embedded/src/Manifest/EmbeddedFilesManifest.cs#L87~~ Targets old TFMs
- [x]  https://github.com/dotnet/aspnetcore/blob/83237c9afaa49a9cb90ceb4da56dddb744baeed0/src/Shared/MediaType/HttpTokenParsingRule.cs#L69
- [x]  https://github.com/dotnet/aspnetcore/blob/83237c9afaa49a9cb90ceb4da56dddb744baeed0/src/Http/Shared/HttpRuleParser.cs#L94 (which looks like a copy of the one in Runtime, so could just copy https://github.com/dotnet/runtime/pull/78660)
- [x]  All the checks in https://github.com/dotnet/aspnetcore/blob/main/src/Middleware/HttpOverrides/src/ForwardedHeadersMiddleware.cs
- [ ]  https://github.com/dotnet/aspnetcore/blob/83237c9afaa49a9cb90ceb4da56dddb744baeed0/src/Http/Headers/src/ContentDispositionHeaderValue.cs#L613 (also already done in runtime in https://github.com/dotnet/runtime/pull/78660)
- [ ]  https://github.com/dotnet/aspnetcore/blob/83237c9afaa49a9cb90ceb4da56dddb744baeed0/src/Http/Shared/CookieHeaderParserShared.cs#L238
- [ ]  https://github.com/dotnet/aspnetcore/blob/83237c9afaa49a9cb90ceb4da56dddb744baeed0/src/Mvc/Mvc.ViewFeatures/src/Rendering/TagBuilder.cs#L428
- [ ] https://github.com/dotnet/aspnetcore/blob/83237c9afaa49a9cb90ceb4da56dddb744baeed0/src/Mvc/Mvc.Razor/src/TagHelpers/UrlResolutionTagHelper.cs#L52

### PR DESCRIPTION
**Pull request by ladeak**
# Using SearchValues in ContentDispositionHeaderValue

Task of https://github.com/dotnet/aspnetcore/issues/46484

## Description

Updating the current Encode5987 method to use SearchValues and IndexOfAny method during the encode of the FileNameStar property. Adding unit tests and a benchmark to measure the before-after performance.

Compared to the .NET runtime [implementation](https://source.dot.net/#System.Net.Http/System/Net/Http/Headers/HeaderUtilities.cs,696f2b0de2ccd5de) a second `Span<char>` is pooled because `ValueStringBuilder` used by the .NET runtime is internal, hence `Encoding.ASCII.GetChars` needs a way to reference the chars passed to the `StringBuilder`

## Performance Comparison:

BenchmarkDotNet=v0.13.0, OS=Windows 10.0.22631
12th Gen Intel Core i7-1255U, 1 CPU, 12 logical and 10 physical cores
.NET SDK=9.0.100-preview.4.24201.1
  [Host]     : .NET 9.0.0 (9.0.24.20403), X64 RyuJIT
  Job-SUANLL : .NET 9.0.0 (9.0.24.18101), X64 RyuJIT

Server=True  Toolchain=.NET Core 9.0  RunStrategy=Throughput

BEFORE:

|                 Method |     Mean |   Error |  StdDev |        Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |---------:|--------:|--------:|------------:|-------:|------:|------:|----------:|
|   FileNameStarEncoding | 213.8 ns | 1.79 ns | 1.68 ns | 4,676,383.4 | 0.0052 |     - |     - |     496 B |
| FileNameStarNoEncoding | 201.0 ns | 1.32 ns | 1.23 ns | 4,975,577.6 | 0.0050 |     - |     - |     464 B |

AFTER **UPDATED: 2024/04/13 corresponding to commit [With custom utf8 and hex encoding combined](https://github.com/dotnet/aspnetcore/pull/55039/commits/2eadb693477f2556fc50787575bc325b3b22c6de)**:

|                 Method |      Mean |    Error |   StdDev |         Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |----------:|---------:|---------:|-------------:|-------:|------:|------:|----------:|
|   FileNameStarEncoding | 213.61 ns | 1.828 ns | 1.710 ns |  4,681,410.7 | 0.0052 |     - |     - |     496 B |
| FileNameStarNoEncoding |  87.12 ns | 0.962 ns | 0.853 ns | 11,478,216.6 | 0.0039 |     - |     - |     368 B |

Fixes #46484

### CODE REVIEW

**Comment by ladeak** [comment]
Adding @amcasey as he has been reviewing the previous related PRs.

**Comment by amcasey** [comment]
To confirm, `FileNameStarEncoding` is _worse_ after the change?

**Comment by ladeak** [comment]
> To confirm, `FileNameStarEncoding` is _worse_ after the change?

`Encoding` means it has characters to actually convert to HexString while `NoEncoding` does not have, but it still does make sure nothing to encode.
The actual results will also greatly depend on the number and frequency of chars to be converted to Hex.

**Comment by amcasey** [comment]
> > To confirm, `FileNameStarEncoding` is _worse_ after the change?
> 
> `Encoding` means it has characters to actually convert to HexString while `NoEncoding` does not have, but it still does make sure nothing to encode. The actual results will also greatly depend on the number and frequency of chars to be converted to Hex.

Sorry, I think we got our lines crossed.  I know encoding is expected to be slower than no-encoding (by a variable amount), but it looks like encoding-before is slower than encoding-after?

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (amcasey) [inline src/Http/Headers/src/ContentDispositionHeaderValue.cs:668]: You didn't change this and it probably isn't worth changing, but I kind of wonder whether it would simplify the code (without hurting perf?) to just escape a single non-ascii character per iteration of the outer loop (rather than computing the length of the non-ascii sequence and processing it in multiple nested loops).

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (amcasey) [inline src/Http/Headers/src/ContentDispositionHeaderValue.cs:643]: My naive guess would be that this is slower than just casting byte to char and that's why the no-encoding case is slightly slower.

**Review by amcasey — COMMENTED** [review]
The change looks generally correct, but I think the possible perf regression in the no-encoding case is worth revisiting.

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (amcasey) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:643]: Do you want to try re-running the benchmarks with this converted to a simple for-loop, casting and appending each byte?

Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:643]: But the "NoEncoding" is actually faster

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:647]: Updating.

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:668]: Yes, I don't think we need this `if` at all. It isn't there for the runtime implementation either, and Rune is only to tell the bytes count, but the old implementation also Hex's a single byte only independent of whatever Rune says.

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (amcasey) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:643]: Oops, I got confused.  Any idea why the encoding part is slower?

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:668]: updated

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:644]: So if I switch this back to a loop, something like:

```csharp
            var asAscii = inputBytes.Slice(0, length);
            while (asAscii.Length > 0)
            {
                builder.Append((char)asAscii[0]);
                asAscii = asAscii.Slice(1);
            }
```

then these would be the results:

```
|                 Method |     Mean |   Error |  StdDev |        Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |---------:|--------:|--------:|------------:|-------:|------:|------:|----------:|
|   FileNameStarEncoding | 217.9 ns | 1.91 ns | 1.69 ns | 4,590,223.2 | 0.0052 |     - |     - |     496 B |
| FileNameStarNoEncoding | 152.4 ns | 1.42 ns | 1.26 ns | 6,561,686.0 | 0.0050 |     - |     - |     464 B |
```

Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:644]: I have just pushed an update where I remove the Rune, and now the "encoding" is more-or less the same considering the error. (I suppose it must be a little slower for having to 'allocate' a second char buffer on stack.

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (ladeak) [inline src/Http/Headers/src/ContentDispositionHeaderValue.cs:628]: This is something that looks off and currently fixing it.

**Comment by ladeak** [comment]
Here is my latest run:

|                 Method |      Mean |    Error |   StdDev |         Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |----------:|---------:|---------:|-------------:|-------:|------:|------:|----------:|
|   FileNameStarEncoding | 211.84 ns | 1.965 ns | 1.838 ns |  4,720,566.7 | 0.0052 |     - |     - |     496 B |
| FileNameStarNoEncoding |  97.99 ns | 0.685 ns | 0.608 ns | 10,205,254.5 | 0.0039 |     - |     - |     368 B |

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (amcasey) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:645]: Yeah, sorry for the confusion about the for-loop - it makes sense that that would be slower.

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (amcasey) [inline src/Http/Headers/src/ContentDispositionHeaderValue.cs:629]: I'm missing something - isn't `input.Length` already in characters (vs bytes)?

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (amcasey) [inline src/Http/Headers/src/ContentDispositionHeaderValue.cs:632]: I'm still curious why this wouldn't be `maxInputAsciiCount`, as opposed to `MaxStackAllocSizeChars`.  Isn't the former smaller?

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (amcasey) [inline src/Http/Headers/src/ContentDispositionHeaderValue.cs:645]: Could we avoid the pool rental by chunking `inputBytes` so that the individual chunks fit within the stackalloc'd buffer?

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:632]: Yes, of course. The copy-paste issue from above.

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:645]: I will do a comparison

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:629]: Yeah the UTF8 conversion confused me, but I don't think we need that line at all as you point out. Actually does more harm then good.

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:645]: I tried avoiding the string -> UTF8 -> ASCII encoding in the case we know it is a limited set of ASCII chars only, but apparently that does not make it faster. I think it could be related to `SearchValues<>` closed generic on `char` vs. `byte`.

I am still following up on the commented suggestion.

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (BrennanConroy) [inline src/Http/Headers/src/ContentDispositionHeaderValue.cs:689]: `builder.Append($"%{HexUpperChars[(b & 0xf0) >> 4]}{HexUpperChars[b & 0xf]}")` is probably in theory faster.

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (BrennanConroy) [inline src/Http/Headers/src/ContentDispositionHeaderValue.cs:663]: Can you try a for loop instead so we don't call Slice repeatedly?

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:644]: So if I understod what you suggest the `tempCharBuffer` allocation changes to:

```csharp
Span<char> tempCharBuffer = stackalloc char[input.Length <= MaxStackAllocSizeChars ? input.Length : MaxStackAllocSizeChars];
```
and the Ascii encoding:

```suggestion
 var toAscii = inputBytes.Slice(0, length);
 while (toAscii.Length > 0)
 {
     var chunkSize = toAscii.Length <= tempCharBuffer.Length ? toAscii.Length : tempCharBuffer.Length;
     int asciiCharCount = Encoding.ASCII.GetChars(toAscii.Slice(0, chunkSize), tempCharBuffer);
     builder.Append(tempCharBuffer[..asciiCharCount]);
     toAscii = toAscii.Slice(chunkSize);
 }
```
My performance results:

|                 Method |     Mean |   Error |  StdDev |        Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |---------:|--------:|--------:|------------:|-------:|------:|------:|----------:|
|   FileNameStarEncoding | 235.9 ns | 2.38 ns | 1.99 ns | 4,239,131.0 | 0.0052 |     - |     - |     496 B |
| FileNameStarNoEncoding | 109.2 ns | 0.53 ns | 0.44 ns | 9,156,655.4 | 0.0039 |     - |     - |     368 B |

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:689]: But it allocates more, no?

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:663]: This seems to have benefited a few ns improvement.

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (BrennanConroy) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:689]: No, it uses the string interpolator overload

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (BrennanConroy) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:644]: No, it would be something like
```
var asciiBuffer = Unsafe.As<byte, char>(inputBytes.Slice(0, length));
builder.Append(asciiBuffer);
```

That code is likely wrong, but it's the general idea.

**Comment by ladeak** [comment]
With for loop to process the hex strings and using string interpolated overload of the string builder, I get the following perf results:


|                 Method |     Mean |   Error |  StdDev |        Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |---------:|--------:|--------:|------------:|-------:|------:|------:|----------:|
|   FileNameStarEncoding | 235.0 ns | 3.15 ns | 2.94 ns | 4,256,174.6 | 0.0052 |     - |     - |     496 B |
| FileNameStarNoEncoding | 101.3 ns | 1.75 ns | 1.63 ns | 9,867,874.7 | 0.0039 |     - |     - |     368 B |

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs]: I implemented posted the results. I could not measure a difference with the perf test.

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs]: It seems on my measurements that using a compile time MaxStackAllocSizeChars has a smaller perf benefit.

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:644]: I am thinking about `MemoryMarshal.Cast<byte, char>(inputBytes.Slice(0, length));`, but I wonder if this will be correct, because UTF-8 bytes uses 1 byte for a character and `char` 2 bytes. So, reinterpreting the byte array would not be correct, would it?

Comment (amcasey) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:644]: It's certainly not intuitively correct, but it's not impossible that there's special handling for this case.

Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:644]: `MemoryMarshal.Cast<byte, char>(inputBytes.Slice(0, length));` got me gibberish, I am looking for alternatives still.

Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:644]: I am looking at an alternative approach promises some benefits, I will polish the code and push it over the weekend

|                 Method |      Mean |    Error |   StdDev |         Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |----------:|---------:|---------:|-------------:|-------:|------:|------:|----------:|
|   FileNameStarEncoding | 224.93 ns | 1.833 ns | 1.715 ns |  4,445,824.9 | 0.0052 |     - |     - |     496 B |
| FileNameStarNoEncoding |  87.70 ns | 0.887 ns | 0.787 ns | 11,402,554.9 | 0.0039 |     - |     - |     368 B |

Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs]: ### Step 1 - commit [Using Rune to encode/decode non ascii characters](https://github.com/dotnet/aspnetcore/pull/55039/commits/7a7f2fd20d64ef551d03fd9ea2d3497783828151)

I have implemented the idea I suggested above. See commit  where the idea is to 
1. For Rfc5987AttrChar append characters to `StringBuilder` directly
1. For non-Rfc5987AttrChars keep using Rune to encode single character-by-character (or 2-chars by 2-chars (Rune)).

The results are:

|                 Method |      Mean |    Error |   StdDev |         Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |----------:|---------:|---------:|-------------:|-------:|------:|------:|----------:|
|   FileNameStarEncoding | 221.70 ns | 1.567 ns | 1.389 ns |  4,510,656.9 | 0.0052 |     - |     - |     496 B |
| FileNameStarNoEncoding |  89.11 ns | 0.905 ns | 0.802 ns | 11,221,787.5 | 0.0039 |     - |     - |     368 B |


### Step 2 - commit [With custom utf8 and hex encoding combined](https://github.com/dotnet/aspnetcore/pull/55039/commits/2eadb693477f2556fc50787575bc325b3b22c6de)

Step 2 takes this a step forward and "borrows" the implementation of [Rune's EncodeToUtf8](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Text/Rune.cs,6487d779e626351e,references) and combines it with the hex creation. This avoids the inner loop around the Hex creation.

My best measured results using this approach:

|                 Method |      Mean |    Error |   StdDev |         Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |----------:|---------:|---------:|-------------:|-------:|------:|------:|----------:|
|   FileNameStarEncoding | 213.61 ns | 1.828 ns | 1.710 ns |  4,681,410.7 | 0.0052 |     - |     - |     496 B |
| FileNameStarNoEncoding |  87.12 ns | 0.962 ns | 0.853 ns | 11,478,216.6 | 0.0039 |     - |     - |     368 B |

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs]: This code segment is removed.

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (BrennanConroy) [inline src/Http/Headers/src/ContentDispositionHeaderValue.cs:662]: Not a fan of copying this code. How much would we lose if we just `stackalloc byte[4]` and called `TryEncodeToUtf8` and looped over the result to write to the `StringBuilder`?

Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:662]: https://github.com/dotnet/aspnetcore/pull/55039#discussion_r1564112587

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (gfoidl) [inline src/Http/Headers/src/ContentDispositionHeaderValue.cs:643]: Nit
```suggestion
            for (var i = 0; i < length;)
```

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (amcasey) [inline src/Http/Headers/src/ContentDispositionHeaderValue.cs:645]: This was probably discussed elsewhere, but why did this change from Utf8 to Utf16?  Was the old code wrong?  Is the new code operating on different data?

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (amcasey) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:662]: Personally, I can live with the copy since it's an implementation of spec'd behavior.  I don't think we'll get out of sync in a way that reduces correctness, but we could well fail to benefit from future perf wins.  If there's a way around copying, that would be preferable, but I wouldn't block the PR for this.  Having said that, I also wouldn't go ahead without @BrennanConroy's approval.

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (amcasey) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:645]: Ah, we did stop calling `Encoding.UTF8.GetBytes` and start operating on `input` directly.  Makes sense.

**Review by amcasey — APPROVED** [review]
LGTM, subject to @BrennanConroy's approval.

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (ladeak) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:645]: Yes, that is removed, hence used Utf16 with the Rune.

Comment (amcasey) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:645]: Yep, LGTM.  I just added the second comment in case anyone else was confused.

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (BrennanConroy) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:662]: That's fair 👍

**Review by BrennanConroy — APPROVED** [review]

**Comment by amcasey** [comment]
Thanks, @ladeak!

**Pull request merged**
